File: 1319.cpp
class Solution {
    vector<int> parent;      
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        if (n - 1 > connections.size()){
            return -1;
        }
        parent.resize(n);
        for(int i = 0; i < n; ++i){ parent[i] = i; }
	    int count = -1;

	    for (int i = 0; i < connections.size(); ++i) {
            if (find(connections[i][0]) != find(connections[i][1])) {
                unite(connections[i][0], connections[i][1]);
            }
        }

        for (int i = 0; i < n; ++i) {
            if (i == parent[i]) {
                count++;
            }
        }
        return count;
    }

    int find(int x){
        if(parent[x] == x) { return x; }
        return parent[x] = find(parent[x]);
    }

    void unite(int x, int y){
        parent[find(x)] = find(y);
    }   
};

File: 1319.py
from typing import List
from collections import deque, defaultdict


class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1

        graph = defaultdict(list)
        for u, v in connections:
            graph[u].append(v)
            graph[v].append(u)

        visited = [False] * n

        def bfs(start):
            queue = deque([start])
            visited[start] = True
            while queue:
                node = queue.popleft()
                for neighbor in graph[node]:
                    if not visited[neighbor]:
                        visited[neighbor] = True
                        queue.append(neighbor)

        components = 0
        for i in range(n):
            if not visited[i]:
                bfs(i)
                components += 1

        return components - 1
